Iportamos las librerias nesesarias

In [1]:
import json
import re
import pandas as pd
import os
import csv

Al tratar de abrir archivo australian_users_items.json posee demasiados errores así que se procede a separar palabras, se cambia  comillas simples por dobles ,se compone las separaciones y gradamos el resultado en un archivó para los datos y otro para los errores que no se pudieron corregir.

In [7]:
entries = []
errores = []

# Función para manejar NaN, Infinity, -Infinity
def handle_nan_constant(value):
    if value in ('NaN', 'Infinity', '-Infinity'):
        return None
    return value

# Función para preprocesar el archivo y corregir el formato JSON
def preprocess_json(data):
    data = data.replace("'", '"')
    data = re.sub(r'}\s*{', '},{', data)
    data = re.sub(r'"(\d+)"', r'\1', data)
    data = re.sub(r'"(true|false)"', r'\1', data)
    data = re.sub(r'\s+', ' ', data)

    return data

# Función para validar que los campos obligatorios existan en un objeto JSON
def validate_json_entry(obj):
    required_keys = ['user_id', 'steam_id', 'user_url']
    for key in required_keys:
        if key not in obj or not obj[key]:
            return False
    return True

# Función para procesar cada objeto del JSON
def process_entry(obj, entry_number):
    try:
        if validate_json_entry(obj):
            user_id = obj.get('user_id', '')
            steam_id = obj.get('steam_id', '')
            user_url = obj.get('user_url', '')
            items_count = obj.get('items_count', 0)
            items = obj.get('items', [])

            # Construimos la lista de ítems
            item_list = []
            for item in items:
                item_id = item.get('item_id', '')
                item_name = item.get('item_name', '')
                playtime_forever = item.get('playtime_forever', 0)
                playtime_2weeks = item.get('playtime_2weeks', 0)
                item_list.append({
                    'item_id': item_id,
                    'item_name': item_name,
                    'playtime_forever': playtime_forever,
                    'playtime_2weeks': playtime_2weeks
                })

            # Agregamos la entrada válida a la lista de entradas
            entries.append({
                'user_id': user_id,
                'steam_id': steam_id,
                'user_url': user_url,
                'items_count': items_count,
                'items': item_list
            })
        else:
            raise ValueError("Faltan campos obligatorios en los datos.")

    except json.JSONDecodeError as json_err:
        # Registrar el error de formato JSON
        print(f'Error de formato en la entrada {entry_number}: {json_err}')
        errores.append({
            'entry_number': entry_number,
            'error': 'JSONDecodeError',
            'message': str(json_err),
            'entry': obj
        })
    except Exception as e:
        # Registrar cualquier otro tipo de error
        print(f'Error al procesar la entrada {entry_number}: {e}')
        errores.append({
            'entry_number': entry_number,
            'error': 'GeneralError',
            'message': str(e),
            'entry': obj
        })

# Cargar el archivo JSON completo
try:
    with open('/content/australian_users_items.json', 'r', encoding='utf-8', errors='ignore') as f:
        data = f.read()
    data = preprocess_json(data)

    print("Archivo cargado y preprocesado correctamente.")

    try:
        json_data = json.loads(data, parse_constant=handle_nan_constant)
        for entry_number, obj in enumerate(json_data, 1):
            process_entry(obj, entry_number)

        # Verificar el tamaño de las listas de entradas y errores
        print(f"Entradas válidas procesadas: {len(entries)}")
        print(f"Errores encontrados: {len(errores)}")

        if entries:
            print("Guardando entradas válidas...")
            with open('australian_users_items_Corrected.json', 'w', encoding='utf-8') as f:
                json.dump(entries, f, ensure_ascii=False, indent=4)
            print("Entradas válidas guardadas en 'australian_users_items_Corrected.json'.")

        # Guardar las líneas con errores en un archivo separado
        if errores:
            print("Guardando entradas con errores...")
            with open('australian_users_items_errors.json', 'w', encoding='utf-8') as f:
                json.dump(errores, f, ensure_ascii=False, indent=4)
            print("Entradas con errores guardadas en 'australian_users_items_errors.json'.")

    except json.JSONDecodeError as e:
        print(f"Error al cargar el archivo JSON completo: {e}")

except FileNotFoundError as e:
    print(f"Archivo no encontrado: {e}")
except Exception as e:
    print(f"Ocurrió un error: {e}")




Archivo cargado y preprocesado correctamente.
Error al cargar el archivo JSON completo: Expecting ',' delimiter: line 1 column 1364 (char 1363)


Después de componer el archivo no pudo componerse línea 1364 al ser poca la información la corregimos manualmente y la sumamos al final

In [8]:
# Datos corregidos manualmente
nuevo_dato = {

  "user_id": "UberDuckLord",
  "items_count": 126,
  "steam_id": "76561198046978928",
  "user_url": "http://steamcommunity.com/id/UberDuckLord",
  "items": [
    {
      "item_id": "220",
      "item_name": "Half-Life 2",
      "playtime_forever": 100,
      "playtime_2weeks": 0
    },
    {
      "item_id": "340",
      "item_name": "Half-Life 2: Lost Coast",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "240",
      "item_name": "Counter-Strike: Source",
      "playtime_forever": 356,
      "playtime_2weeks": 0
    },
    {
      "item_id": "4000",
      "item_name": "Garry's Mod",
      "playtime_forever": 115412,
      "playtime_2weeks": 0
    },
    {
      "item_id": "2400",
      "item_name": "The Ship",
      "playtime_forever": 73,
      "playtime_2weeks": 0
    },
    {
      "item_id": "2420",
      "item_name": "The Ship Single Player",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "2430",
      "item_name": "The Ship Tutorial",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "17390",
      "item_name": "Spore",
      "playtime_forever": 974,
      "playtime_2weeks": 0
    },
    {
      "item_id": "1250",
      "item_name": "Killing Floor",
      "playtime_forever": 76,
      "playtime_2weeks": 0
    },
    {
      "item_id": "35420",
      "item_name": "Killing Floor Mod: Defence Alliance 2",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "550",
      "item_name": "Left 4 Dead 2",
      "playtime_forever": 4757,
      "playtime_2weeks": 0
    },
    {
      "item_id": "223530",
      "item_name": "Left 4 Dead 2 Beta",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "8190",
      "item_name": "Just Cause 2",
      "playtime_forever": 112,
      "playtime_2weeks": 0
    },
    {
      "item_id": "33910",
      "item_name": "Arma 2",
      "playtime_forever": 5,
      "playtime_2weeks": 0
    },
    {
      "item_id": "33930",
      "item_name": "Arma 2: Operation Arrowhead",
      "playtime_forever": 97,
      "playtime_2weeks": 0
    },
    {
      "item_id": "219540",
      "item_name": "Arma 2: Operation Arrowhead Beta (Obsolete)",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "31280",
      "item_name": "Poker Night at the Inventory",
      "playtime_forever": 3,
      "playtime_2weeks": 0
    },
    {
      "item_id": "40800",
      "item_name": "Super Meat Boy",
      "playtime_forever": 39,
      "playtime_2weeks": 0
    },
    {
      "item_id": "42910",
      "item_name": "Magicka",
      "playtime_forever": 2702,
      "playtime_2weeks": 0
    },
    {
      "item_id": "91600",
      "item_name": "Sanctum",
      "playtime_forever": 532,
      "playtime_2weeks": 0
    },
    {
      "item_id": "620",
      "item_name": "Portal 2",
      "playtime_forever": 134,
      "playtime_2weeks": 0
    },
    {
      "item_id": "105600",
      "item_name": "Terraria",
      "playtime_forever": 7445,
      "playtime_2weeks": 1
    },
    {
      "item_id": "24240",
      "item_name": "PAYDAY: The Heist",
      "playtime_forever": 549,
      "playtime_2weeks": 0
    },
    {
      "item_id": "113200",
      "item_name": "The Binding of Isaac",
      "playtime_forever": 1442,
      "playtime_2weeks": 0
    },
    {
      "item_id": "201870",
      "item_name": "Assassin's Creed Revelations",
      "playtime_forever": 86,
      "playtime_2weeks": 0
    },
    {
      "item_id": "72850",
      "item_name": "The Elder Scrolls V: Skyrim",
      "playtime_forever": 5469,
      "playtime_2weeks": 0
    },
    {
      "item_id": "22380",
      "item_name": "Fallout: New Vegas",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "209080",
      "item_name": "Guns of Icarus Online",
      "playtime_forever": 350,
      "playtime_2weeks": 0
    },
    {
      "item_id": "730",
      "item_name": "Counter-Strike: Global Offensive",
      "playtime_forever": 13010,
      "playtime_2weeks": 0
    },
    {
      "item_id": "204300",
      "item_name": "Awesomenauts",
      "playtime_forever": 93,
      "playtime_2weeks": 0
    },
    {
      "item_id": "214560",
      "item_name": "Mark of the Ninja",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "204360",
      "item_name": "Castle Crashers",
      "playtime_forever": 148,
      "playtime_2weeks": 0
    },
    {
      "item_id": "200710",
      "item_name": "Torchlight II",
      "playtime_forever": 81,
      "playtime_2weeks": 0
    },
    {
      "item_id": "205790",
      "item_name": "Dota 2 Test",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "219640",
      "item_name": "Chivalry: Medieval Warfare",
      "playtime_forever": 349,
      "playtime_2weeks": 0
    },
    {
      "item_id": "232210",
      "item_name": "Patch testing for Chivalry",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "223470",
      "item_name": "POSTAL 2",
      "playtime_forever": 441,
      "playtime_2weeks": 0
    },
    {
      "item_id": "208480",
      "item_name": "Assassin's Creed® III",
      "playtime_forever": 1016,
      "playtime_2weeks": 0
    },
    {
      "item_id": "202970",
      "item_name": "Call of Duty: Black Ops II",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "202990",
      "item_name": "Call of Duty: Black Ops II - Multiplayer",
      "playtime_forever": 0,
      "playtime_2weeks": 0
    },
    {
      "item_id": "212910",
      "item_name": "Call of Duty: Black Ops II - Zombies",
      "playtime_forever": 2836,
      "playtime_2weeks": 0
    },
    {
      "item_id": "224540",
      "item_name": "Ace of Spades",
      "playtime_forever": 516,
      "playtime_2weeks": 0
    },
    {
      "item_id": "204080",
      "item_name": "The Showdown Effect",
      "playtime_forever": 0,
      "playtime_2weeks": 0

        }
    ]
}

# Ruta al archivo JSON
ruta_archivo = '/content/australian_users_items_Corrected .json'

# Verificar si el archivo ya existe
if os.path.exists(ruta_archivo):
    print(f"El archivo {ruta_archivo} existe, cargando datos...")
else:
    print(f"El archivo {ruta_archivo} no existe, se creará uno nuevo.")

# Cargar datos existentes del archivo JSON
try:
    with open(ruta_archivo, 'r') as archivo:
        datos_existentes = json.load(archivo)
        print("Datos existentes cargados exitosamente.")
except FileNotFoundError:
    datos_existentes = []
    print(f"Archivo no encontrado. Se creará un nuevo archivo en {ruta_archivo}.")
except json.JSONDecodeError:
    datos_existentes = []
    print(f"Archivo vacío o malformado. Iniciando con una lista vacía.")

# Agregar el nuevo dato al contenido existente
datos_existentes.append(nuevo_dato)

# Guardar los datos actualizados en el archivo JSON
try:
    with open(ruta_archivo, 'w') as archivo:
        json.dump(datos_existentes, archivo, indent=4)
        print("Datos guardados exitosamente.")
except IOError as e:
    print(f"Error al guardar los datos: {e}")

El archivo /content/australian_users_items_Corrected .json existe, cargando datos...
Datos existentes cargados exitosamente.
Datos guardados exitosamente.


Una ves que el archivo  a sido reparado lo pasaremos a formato csv donde cambiamos los nombres para una mejor comprensión  ademar convertimos los minutos a horas, de existir errores lo guardaremos en un nuevo archivo

In [9]:
import json
import csv

# Ruta de los archivos
json_file_path = '/content/australian_users_items_Corrected .json'  # Corregir espacio extra en el nombre
csv_output_path = './australian_users_items.csv'  # Ruta del archivo CSV de salida

# Función para convertir minutos a horas
def convertir_tiempo(minutos):
    if minutos == 0:
        return "NaN"  # Usar "NaN" como cadena para valores sin tiempo jugado
    horas = minutos // 60
    return f"{horas}h" if horas > 0 else "NaN"

# Función para procesar y escribir cada entrada en el archivo CSV
def procesar_entrada_usuario(usuario, escritor_csv):
    id_usuario = usuario.get('user_id', 'NaN')
    id_steam = usuario.get('steam_id', 'NaN')
    url_usuario = usuario.get('user_url', 'NaN')
    numero_juegos = usuario.get('items_count', 'NaN')

    # Procesar los juegos en la lista "items"
    for item in usuario.get('items', []):
        id_juego = item.get('item_id', 'NaN')
        nombre_juego = item.get('item_name', 'NaN')
        tiempo_jugado_total = item.get('playtime_forever', 0)
        tiempo_jugado_semanas = item.get('playtime_2weeks', 0)

        # Convertir los tiempos a un formato amigable
        tiempo_jugado_total_amigable = convertir_tiempo(tiempo_jugado_total)
        tiempo_jugado_semanas_amigable = convertir_tiempo(tiempo_jugado_semanas)

        # Escribir los datos en el CSV
        escritor_csv.writerow({
            'Id_usuario': id_usuario,
            'Id_Steam': id_steam,
            'Url_usuario': url_usuario,
            'Numero_de_juegos': numero_juegos,
            'Id_juego': id_juego,
            'Nombre_del_juego': nombre_juego,
            'Tiempo_jugado_total': tiempo_jugado_total_amigable,
            'Tiempo_jugado_en_2_semanas': tiempo_jugado_semanas_amigable
        })

# Función para procesar todo el archivo JSON y escribirlo en CSV
def procesar_archivo_json(json_file_path, csv_output_path):
    try:
        # Cargar el archivo JSON
        with open(json_file_path, 'r', encoding='utf-8') as json_file:
            datos = json.load(json_file)

        # Definir los encabezados del CSV
        campos = ['Id_usuario', 'Id_Steam', 'Url_usuario', 'Numero_de_juegos', 'Id_juego', 'Nombre_del_juego', 'Tiempo_jugado_total', 'Tiempo_jugado_en_2_semanas']

        # Abrir el archivo CSV para escribir
        with open(csv_output_path, 'w', newline='', encoding='utf-8') as csv_file:
            escritor_csv = csv.DictWriter(csv_file, fieldnames=campos)
            escritor_csv.writeheader()  # Escribir la fila de encabezados

            # Procesar cada usuario en el JSON
            for usuario in datos:
                procesar_entrada_usuario(usuario, escritor_csv)

        print(f"Archivo CSV generado correctamente en '{csv_output_path}'.")

    except FileNotFoundError:
        print(f"Error: El archivo '{json_file_path}' no fue encontrado.")
    except json.JSONDecodeError:
        print(f"Error: No se pudo decodificar el archivo JSON '{json_file_path}'.")
    except Exception as e:
        print(f"Error inesperado: {e}")

# Llamar a la función para procesar el archivo JSON y generar el CSV
procesar_archivo_json(json_file_path, csv_output_path)


Archivo CSV generado correctamente en './australian_users_items.csv'.


EDA

Al no haber errores en el archivo csv  comenzamos el EDA cargando  la información en un data frame y visualizamos las primeras líneas

In [11]:
file_path = '/content/australian_users_items.csv'

df = pd.read_csv(file_path)

df.head()

Id_usuario           Id_Steam                                Url_usuario  \
0  UberDuckLord  76561198046978928  http://steamcommunity.com/id/UberDuckLord   
1  UberDuckLord  76561198046978928  http://steamcommunity.com/id/UberDuckLord   
2  UberDuckLord  76561198046978928  http://steamcommunity.com/id/UberDuckLord   
3  UberDuckLord  76561198046978928  http://steamcommunity.com/id/UberDuckLord   
4  UberDuckLord  76561198046978928  http://steamcommunity.com/id/UberDuckLord   

   Numero_de_juegos  Id_juego         Nombre_del_juego Tiempo_jugado_total  \
0               126       220              Half-Life 2                  1h   
1               126       340  Half-Life 2: Lost Coast                 NaN   
2               126       240   Counter-Strike: Source                  5h   
3               126      4000              Garry's Mod               1923h   
4               126      2400                 The Ship                  1h   

   Tiempo_jugado_en_2_semanas  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN

Realizamos un resumen estadístico delos datos numéricos, hacemos un resumen de cantidad de filas y columnas , obtenemos información del archivo y revisamos la cantidad de datos nulos por columna

In [12]:
df.describe(include='object')
print(f"Dimensiones del DataFrame: {df.shape}")

df.info()

print("Valores nulos por columna:")
print(df.isnull().sum())

Dimensiones del DataFrame: (86, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86 entries, 0 to 85
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Id_usuario                  86 non-null     object 
 1   Id_Steam                    86 non-null     int64  
 2   Url_usuario                 86 non-null     object 
 3   Numero_de_juegos            86 non-null     int64  
 4   Id_juego                    86 non-null     int64  
 5   Nombre_del_juego            86 non-null     object 
 6   Tiempo_jugado_total         54 non-null     object 
 7   Tiempo_jugado_en_2_semanas  0 non-null      float64
dtypes: float64(1), int64(3), object(4)
memory usage: 5.5+ KB
Valores nulos por columna:
Id_usuario                     0
Id_Steam                       0
Url_usuario                    0
Numero_de_juegos               0
Id_juego                       0
Nombre_del_juego               0
Tiempo